Remove Python 2 workarounds

terrdavis opened the pull request:
I don't know if docstring changes belong in a different PR.
I'd be happy to split this one up if needed.
This also includes one or two things related to python<3.5.

One of the docstring changes I made included a note about using a context manager. Could I replace the begin/end_group pairs with a context manager in this PR?
```python
    def begin_group(self, indent=0, open=''):	
        """
        Begin a group.  If you want support for python < 2.5 which doesn't has	        Begin a group.
        the with statement this is the preferred way:	
            p.begin_group(1, '{')	
            ...	
            p.end_group(1, '}')	
        The python 2.5 expression would be this:	
            with p.group(1, '{', '}'):	
                ...	
        The first parameter specifies the indentation for the next line (usually	        The first parameter specifies the indentation for the next line (usually
        the width of the opening text), the second the opening text.  All	        the width of the opening text), the second the opening text.  All
        parameters are optional.
```

I also found this test, with a TODO related to python 3.6+ support. I'm not sure what should be removed.
```python
def test_instantiation_FileLink():
    """FileLink: Test class can be instantiated"""
    fl = display.FileLink('example.txt')
    # TODO: remove if when only Python >= 3.6 is supported
    fl = display.FileLink(pathlib.PurePath('example.txt'))
```

terrdavis committed:
Remove clearly marked python 2 imports.

terrdavis committed:
Remove cProfile fallback import, since it was added in Python2.5

terrdavis committed:
Remove None check for fallback cProfile import and ExecutionMagics.profile_missing_notice.

This is the fallback for the fallback case of python < 2.5

terrdavis committed:
Remove sqlite3 fallback imports and associated checks.

sqlite3 was add to the stdlib in python2.5

terrdavis committed:
Remove unused import of sys.

Carreau commented on IPython/core/history.py:
```diff
@@ -8,13 +8,7 @@
-        from pysqlite2 import dbapi2 as sqlite3
-    except ImportError:
-        sqlite3 = None
+import sqlite3
```
I think that not quite right; this make sqlite3 mandatory, before it would set slite3 to None, and still work.

terrdavis replied:
sqlite3 was [added in python 2.5](https://docs.python.org/2/library/sqlite3.html), so I assumed this import would never fail.

Carreau replied:
That seem to have been added in 2012; (#2718), so my guess is some people have-had custom builds that do not have sqlite3  for whatever reason. 

It's been 6 years so let's see if this is still necessary. If we have complaints we can re-add.

Carreau commented on IPython/core/magics/execution.py:
```diff
@@ -19,16 +19,8 @@
-    except ImportError:
-        profile = pstats = None
+import cProfile as profile
+import pstats
```
same here `# profile isn't bundled by default in Debian for license reasons` so this code is necessary for debian.

terrdavis replied:
I saw that, but I assumed that the cProfile import couldn't fail since it was added in 2.5. Is the cProfile import ever expected to fail?

Carreau replied:
As far as I can tell debian used to build python w/o some modules;  it seem to not be the case anymore. But yes `import profile` used to be able to fail.

Carreau commented:
Thanks for doing this ! 

It looks like the test failures are genuine; but in principle from what I have read I do not have any objections to the code change. 

I guess it would be easier in smaller chunks to figure out what is failing.

In general if you want to change some things to be context managers; I do not have objections.

terrdavis commented:
You're welcome!  

I'm confused though, do I need revert the changes that you commented on?  

Can you also clarify what's meant by the comment in this test?
```python
def test_instantiation_FileLink():
    """FileLink: Test class can be instantiated"""
    fl = display.FileLink('example.txt')
    # TODO: remove if when only Python >= 3.6 is supported
    fl = display.FileLink(pathlib.PurePath('example.txt'))
```

Otherwise, I'll split this up into more pull requests and add the context manager I asked about!

Carreau commented:
> do I need revert the changes that you commented on

Maybe not revert, but the changes are going too far, for examplein 

```diff
-try:
-    import sqlite3
-except ImportError:
-    try:
-        from pysqlite2 import dbapi2 as sqlite3
-   except ImportError:
-        sqlite3 = None
+ import sqlite3
```
It probably should be more like: 

```diff
try:
    import sqlite3
except ImportError:
-    try:
-         from pysqlite2 import dbapi2 as sqlite3
-     except ImportError:
        sqlite3 = None
import sqlite3
```

Where we only remove the Py3 specific versions and keep the `@needs_sqlite` which is specific for debian.

Carreau commented:
```
# TODO: remove if when only Python >= 3.6 is supported
fl = display.FileLink(pathlib.PurePath('example.txt'))
```

I hav to check, but that likely mean that one of these two lines is  unnecessary now that we're Python 3.6+ only.

terrdavis commented:
Apparently, cProfile is [included in Debian](https://wiki.debian.org/Python#Deviations_from_upstream):
> The standard profile and pstats modules are not included, due to restrictive redistribution clauses in their license that don’t meet the DFSG. **cProfile and hotshot are however included.**

Is there any other case where cProfile isn't available?

terrdavis commented:
@Carreau, can you respond to my last message? As far as I can tell, cProfile and sqlite3 are included in Debian’s python 3.

Carreau commented:
Sorry, I'll have a look. It might have been an old policy in Debian, in which case we can revisit it, and I'm ok to have those import unconditionally present.

Commit:
revert back ultratb to version on master

Stack_data has been introduced in a previous PRs so woudl have a lot of
merge conflicts.

Carreau commented on IPython/core/history.py:
```diff
@@ -49,26 +43,8 @@ def __exit__(self, *args, **kwargs):
-    class OperationalError(Exception):
-        "Dummy exception when sqlite could not be imported. Should never occur."
+DatabaseError = sqlite3.DatabaseError
+OperationalError = sqlite3.OperationalError
```
This create actual test failures;  The decorator has the following statement:

```
if sqlite3 is None or not self.enabled:
```

Note the `or not self.enabled`; So this decorator is not only about sqlite being available.

Carreau commented:
I've pushed a commit that fix a merge conflict.

terrdavis committed:
Remove sqlite3 None checks and simplify.

terrdavis committed:
Rename decorator needs_sqlite to only_when_enabled.

terrdavis committed:
test_magic: Remove skip_without('sqlite3').

terrdavis committed:
Assume sqlite3 always available in testing.

terrdavis committed:
Remove sqlite3 checks in test setup.

terrdavis committed:
remove broken test.

Got error: "IPython.core.error.UsageError: unrecognized arguments: -rl"
`-r` doesn't appear to be a valid option for %history/%hist.

terrdavis committed:
remove __future__ import of print function

terrdavis commented:
I meant to finish off these changes on a new branch (i.e. remove-py2-workarounds-2of2) but my local branch still had this as its remote. I rebased on to master in the meantime. Let me know if you'd prefer that I close this PR and move it to a new one.

Carreau commented:
> I meant to finish off these changes on a new branch (i.e. remove-py2-workarounds-2of2) but my local branch still had this as its remote. I rebased on to master in the meantime. Let me know if you'd prefer that I close this PR and move it to a new one.

Sorry I've been super busy recently. 

I think something got messed up in the rebase. I'll have a look tomorrow but it should be fixable. If not we can do another PR but that should not be necessary.

Carreau commented:
Note to self, current head is e507997163fc97cf46fc3005331ba5d9f1aeb992

Carreau commented:
Ok, let's try this rebase – I kept thinking about it so had to do it tonight – 13 commits instead of 54.
And it looks good, let's let CI do its thing, and if it pass i'll merge tomorrow.

Carreau commented:
... Or later tonight depending as to whether I'm sleeping or not.

Carreau commented:
Woooo
